connector/src/main/java/com/google/cloud/bigquery/connector/common/IteratorMultiplexer.java

```diff
@@ -0,0 +1,137 @@
+   *
+   * @param iterator The Iterator to read from.
+   * @param splits The number of output iterators that will read from iterator.
+   */
+  IteratorMultiplexer(Iterator<T> iterator, int splits) {
+    this.iterator = iterator;
+    this.splits = splits;
+    // Filled in when initializing iterators.
+    iterators = new QueueIterator[splits];
+    for (int x = 0; x < splits; x++) {
+      iterators[x] = new QueueIterator<>();
+    }
+  }
+
+  @Override
+  public void close() {
+    if (worker != null) {
+      worker.interrupt();
+      try {
+        worker.join(/*millis=*/ 1000);
+      } catch (InterruptedException e) {
+        throw new RuntimeException("Interrupted while waiting on worker thread shutdown.", e);
+      }
+      worker = null;
+      if (rethrow != null) {
+        throw new RuntimeException("Error while closing.", rethrow);
+      }
+    } else {
+      for (int x = 0; x < splits; x++) {
+        iterators[x].done.set(true);
+      }
+    }
+  }
+
+  void readAhead() {
+    try {
+      boolean hasMore = true;
+      while (hasMore) {
+        for (int x = 0; x < splits; x++) {
+          if (iterator.hasNext()) {
+            T value = iterator.next();
+            iterators[x].queue.put(value);
+          } else {
+            hasMore = false;
+            break;
+          }
+        }
+      }
+    } catch (InterruptedException e) {
+      log.info("Worker thread had error. Ending all iterators");
+      rethrow = new RuntimeException("Worker thread interrupted");
+    } catch (RuntimeException e) {
+      rethrow = e;
+    }
+    for (int x = 0; x < splits; x++) {
+      iterators[x].done.set(true);
+    }
+  }
+
+  public Iterator<T> getSplit(int split) {
+    if (worker == null) {
+      worker = new Thread(this::readAhead, "readahead-worker");
+      worker.setDaemon(true);
+      worker.start();
+    }
+    return iterators[split];
+  }
+
+  private class QueueIterator<T> implements Iterator<T> {
+    private final ArrayBlockingQueue<T> queue = new ArrayBlockingQueue<>(/*capacity=*/ 1);
+    private final AtomicBoolean done = new AtomicBoolean(false);
+
+    private T t = null;
+
+    @Override
+    public boolean hasNext() {
+      if (!mightHaveNext()) {
+        return false;
+      }
+      t = null;
+      try {
+        while (t == null && mightHaveNext()) {
+          t = queue.poll(10, TimeUnit.MILLISECONDS);
+        }
+      } catch (InterruptedException e) {
+        done.set(true);
+        // We expect all iterators to either make progress together or finish.
+        // This starts the cleanup process to halt all workers.
+        worker.interrupt();
+        throw new RuntimeException(e);
+      }
+      if (t == null) {
+        done.set(true);
+        return false;
+      }
+      return true;
+    }
+
+    private boolean mightHaveNext() {
+      return !done.get() || !queue.isEmpty();
+    }
+
+    @Override
+    public T next() {
+      Preconditions.checkState(t != null);
```

Comment:
Please add error message "next element cannot be null"

Reply:
done.